runtime/strategies/host-particle.js

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2018 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {Strategy} from '../../strategizer/strategizer.js';
+import {Recipe} from '../recipe/recipe.js';
+import {Walker} from '../recipe/walker.js';
+import {TypeChecker} from '../recipe/type-checker.js';
+import {assert} from '../../platform/assert-web.js';
+
+export class HostParticle extends Strategy {
+
+  constructor(arc) {
+    super();
+    this._arc = arc;
+  }
+
+  async generate(inputParams) {
+    let arc = this._arc;
+    return Recipe.over(this.getResults(inputParams), new class extends Walker {
+      onHandleConnection(recipe, connection) {
+        if (connection.direction !== 'host' || connection.handle) return;
+        assert(connection.type.isInterface);
+
+        let results = [];
+        for (let particle of arc.context.particles) {
+          if (!connection.type.interfaceShape.particleMatches(particle)) continue;
+
+          results.push((recipe, hc) => {
+            // Restricting the type of the connection to the concrete particle
+            // may restrict type variable across the recipe.
+            hc.type.interfaceShape.restrictType(particle);
+
+            // The connection type may still have type variables:
+            // E.g. if shape requires `in ~a *`
+            //      and particle has `in Entity input`
+            //      then type system has to ensure ~a is at least Entity.
+            // The type of a handle hosting the particle literal has to be
+            // concrete, so we concretize connection type with maybeEnsureResolved().
+            let handleType = hc.type.clone(new Map());
+            if (!handleType.maybeEnsureResolved()) return null;
```

Comment:
probably if the type cannot be ensure-resolved, we shouldn't be pushing this particle into results, should we?

Reply:
this is a type of a shape after restricting the type to a particle. Even if particle matches and is fully concrete, the type of the shape can still contain type variables.

E.g. for a shape
```
shape HostedParticleShape
  in ~a *
```

if you try to substitute:
```
particle P
  in Thing input
```

Then ~a stays as a type variable, but has a `canWriteSuperset` constraint of `Thing` - which is correct.

With `maybeEnsureResolved()` I just ensure that the type of this `copy` handle is concrete version of this shape, which I think is ok. WDYT?

Reply:
then in which case the `maybeEnsureResolved` here will return false? 
(i'm sorry, i think i'm somewhat lagging behind in understanding the type system)

Reply:
`maybeEnsureResolved()` would return false if particle being hosted has type variables itself and ~a cannot be resolved to anything. I've refactored this code to move this check up, so that here it cannot fail. I've also added a test _documenting_ the current behavior and a big TODO to figure it out. I will file an issue in a second as well to explore this.